### DISCUSSION TITLE
Correct usage of HasSubstructMatch with SMILES/SMARTS wildcard

### DISCUSSION THREAD

**Discussion by colinrsmall** [rdkit/rdkit#5480] [Q&A] [answered]
I have a set of SMILES string patterns that I'm matching to a larger set of molecules. I'm having issues where certain patterns with wildcards (*) aren't being matched in known control cases:

![image](https://user-images.githubusercontent.com/20507041/182246465-67a92970-b644-4057-95bb-1806ab7d88ca.png)

Removing the wildcard from the string, HasSubstructMatch is able to find the pattern:

![image](https://user-images.githubusercontent.com/20507041/182246609-f073afed-78d8-4ff4-b6e5-7a6e79ca1d7d.png)

This is less of an issue with the pattern above, but inclusion of the wildcard is important for others:

![image](https://user-images.githubusercontent.com/20507041/182246743-079fe542-eec4-422e-b25a-dfc2059354d6.png)

![image](https://user-images.githubusercontent.com/20507041/182246780-8a94b49d-3213-4cc4-8db5-dad6ff970cd3.png)

I read through https://github.com/rdkit/rdkit/issues/3117, but it wasn't explained how wildcards are treated in HasSubstructMatch.

**Comment by DavidACosgrove** [comment]
I expect you’d do better if you created your query with MolFromSmarts.
Generally, queries are SMARTS and molecules are SMILES.


On Mon, 1 Aug 2022 at 22:18, colinrsmall ***@***.***> wrote:

> I have a set of SMILES string patterns that I'm matching to a larger set
> of molecules. I'm having issues where certain patterns with wildcards (*)
> aren't being matched in known control cases:
>
> [image: image]
> <https://user-images.githubusercontent.com/20507041/182246465-67a92970-b644-4057-95bb-1806ab7d88ca.png>
>
> Removing the wildcard from the string, HasSubstructMatch is able to find
> the pattern:
>
> [image: image]
> <https://user-images.githubusercontent.com/20507041/182246609-f073afed-78d8-4ff4-b6e5-7a6e79ca1d7d.png>
>
> This is less of an issue with the pattern above, but inclusion of the
> wildcard is important for others:
>
> [image: image]
> <https://user-images.githubusercontent.com/20507041/182246743-079fe542-eec4-422e-b25a-dfc2059354d6.png>
>
> [image: image]
> <https://user-images.githubusercontent.com/20507041/182246780-8a94b49d-3213-4cc4-8db5-dad6ff970cd3.png>
>
> I read through #3117 <https://github.com/rdkit/rdkit/issues/3117>, but it
> wasn't explained how wildcards are treated in HasSubstructMatch.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/5480>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FXJX2OOB22EPTOREOTVXA5JJANCNFSM55I5EBDA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by colinrsmall** [reply]
Right, that's what I was originally using but ran into more issues with loading the patterns as SMARTS strings, due to (I think) aromaticity specified in the SMARTS string and the presence/lack of explicit hydrogens in the pattern SMARTS strings that we need to match over. Neither pattern matches if loaded as SMARTS whether I include the wildcard or not:

![image](https://user-images.githubusercontent.com/20507041/182257561-8eb864a5-7bf0-4a28-b49f-9b406fd743d8.png)

![image](https://user-images.githubusercontent.com/20507041/182257594-4e3d5b3f-1583-4af7-9934-59f5722e8062.png)

**Reply by DavidACosgrove** [reply]
Your SMARTS strings in these cases don't have aromatic bonds in them, they have a mixture of double bonds and default bonds.  In SMARTS, the default bond is single or aromatic, which is what the pictures are showing.  The double bonds in your query won't match the aromatic bonds in your test molecules.  I think you need to spend some time getting to understand the SMARTS language, which is different from SMILES in a number of important ways.  The definitive guide is https://www.daylight.com/dayhtml_tutorials/languages/smarts/.

**Reply by colinrsmall** [reply]
Thanks. I actually looked into that with someone more trained in chemistry than I am (I'm not a chemist) yesterday, but we made little progress. Took another read through today and I am starting to understand the language ("Nc1nonc1*" would be the right SMARTS query for the first fragment shown to match the reference molecule, although we're confused about the true aromaticity of the oxygen in such a fragment). I think the reason the patterns work when loaded as SMILES strings is that RDKIT infers aromaticity automatically for SMILES (https://sourceforge.net/p/rdkit/mailman/message/23801106/).

This still doesn't explain why the wildcard doesn't work when patterns are loaded as SMILES. I understand that the wildcard is not a part of the base SMILES language, but RDKIT correctly parses and renders the fragment with the wildcard. Perhaps when loaded as a SMILES string, the wildcard isn't treated as a true wildcard in the sense that it won't match any atom or group but is still considered part of the fragment?

We're working with a set of ~500 query patterns (in the form shown in the original images - with the explicit double bonds) that I'm not the original author for, so I'd like to avoid having to manually re-write each pattern. Perhaps an automated solution would be to load each pattern as both a SMILES string (to determine aromaticity) and as a SMARTS string (to be used for matching), identify the aromatic atoms in the SMILES-loaded fragment (with rdqueries.IsAromaticQueryAtom() or something similar), then set the corresponding atoms in the SMARTS-loaded fragment to be aromatic.

**Reply by DavidACosgrove** [reply]
A star in a SMILES isn't a wildcard, it's a dummy atom, with atomic number 0, so that's what you get when you load the string with MolToSmiles.  When used in a query, that atom will match other dummy atoms.  When loaded with MolToSmarts, it will be created as a query atom that matches anything.  You could use MolToSmiles and then loop through the atoms looking for those with atomic number 0, and setting the atom's query properties but that's not something I've ever done so can't offer a bit of code.

**Comment by ptosco** [comment, marked as answer]
@colinrsmall There is a simple way to convert the SMILES dummies into queries:
```
from rdkit import Chem

h = Chem.MolFromSmiles("*CC1=NC=CC=C1")
m = Chem.MolFromSmiles("Cc1ccc(-c2ncccc2OCc2ccccn2)cc1")
m.HasSubstructMatch(h)
False
# Convert dummies into queries
p = Chem.AdjustQueryParameters.NoAdjustments()
p.makeDummiesQueries = True
hq = Chem.AdjustQueryProperties(h, p)
m.HasSubstructMatch(hq)
True
```
Side note: pasting Jupyter Notebook cells as images requires people who wish to help to re-type SMARTS and SMILES patterns, which is time-consuming and error prone. Pasting them as triple-quoted text would make things simpler.

**Reply by colinrsmall** [reply]
Many thanks. Noted on the images.
